src/prototype_room_mineral.js

```diff
@@ -221,3 +221,89 @@ Room.prototype.handleTerminal = function() {
   delete this.memory.mineralOrder[roomOtherName];
   return true;
 };
+
+Room.prototype.cleanupNotEnoughResources = function() {
+  if (!this.memory.reaction) {
+    return false;
+  }
+  const lab1 = Game.getObjectById(this.memory.reaction.labs[1]);
+  const lab2 = Game.getObjectById(this.memory.reaction.labs[2]);
+  const findCreep = this.findMyCreepsOfRole('mineral');
+  const mineralCreep = findCreep.length ? findCreep[0] : false;
+  if (mineralCreep && this.memory.reaction && this.memory.reaction.result) {
+    if (!mineralCreep.checkLabEnoughMineral(lab1, this.memory.reaction.result.first) || !mineralCreep.checkLabEnoughMineral(lab2, this.memory.reaction.result.second)) {
+      roles.mineral.cleanUpLabs(mineralCreep);
+    }
+  }
+  return true;
+};
+
+Room.prototype.handleReaction = function() {
+  if (!this.memory.reaction) {
+    return false;
+  }
+
+  const labsToReact = this.findStructuresOfStructureType(STRUCTURE_LAB).filter((lab) => {
+    if (lab.store[this.memory.reaction.result.first] > 0 || lab.store[this.memory.reaction.result.second] > 0) {
+      return false;
+    }
+    return true;
+  });
+
+  if (labsToReact.length === 0) {
+    return false;
+  }
+
+  const lab0 = Game.getObjectById(this.memory.reaction.labs[0]);
+  const lab1 = Game.getObjectById(this.memory.reaction.labs[1]);
+  const lab2 = Game.getObjectById(this.memory.reaction.labs[2]);
+  if (lab0 === null || lab1 === null || lab2 === null) {
+    delete this.memory.reaction;
+    return false;
+  } else {
```

Comment:
```suggestion
  }
```
(move the `}` up)
The `if` clause ends with a `return`. It is a bit easier to read.